### PR TITLE
cli: streamline autostart ux

### PR DESCRIPTION
**Pull request by johnstcn**
This PR makes the following changes:

- `autostart enable|disable` => `autostart set|unset`
- `autostart enable` now accepts a more natual schedule format: `<time> <days-of-week> <location>`
- `autostart show` now shows configured timezone
- 🎉  automatic timezone detection across mac, windows, linux 🎉

Fixes https://github.com/coder/coder/issues/1647

### CODE REVIEW

File: coderd/util/tz/tz_darwin.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	}
+
+	stripped := strings.Replace(lp, realZoneInfoPath, "", -1)
+	stripped = strings.TrimPrefix(stripped, string(filepath.Separator))
+	loc, err := time.LoadLocation(stripped)
+	if err != nil {
+		return nil, xerrors.Errorf("invalid location %q guessed from %s: %w", stripped, lp, err)
+	}
+	return loc, nil
+}
```
Comment (johnstcn) [inline coderd/util/tz/tz_darwin.go:54]: **Review:** at a glance, the Darwin and Linux versions of these functions seems identical. Be assured they are not.

File: coderd/util/tz/tz_windows.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	}
+	// What we want is the second line of output
+	locStr := strings.TrimSpace(outLines[1])
+	loc, err := time.LoadLocation(locStr)
+	if err != nil {
+		return nil, xerrors.Errorf("invalid location %q from powershell: %w", locStr, err)
+	}
+
+	return loc, nil
+}
```
Comment (johnstcn) [inline coderd/util/tz/tz_windows.go:71]: **Review:** just don't stare too closely at this

File: coderd/autobuild/schedule/schedule.go [hunk 3]
```diff
@@ -74,7 +75,8 @@ func Weekly(raw string) (*Schedule, error) {
 }
 
 // Schedule represents a cron schedule.
-// It's essentially a thin wrapper for robfig/cron/v3 that implements Stringer.
```
Comment (johnstcn) [inline coderd/autobuild/schedule/schedule.go:77]: **Review:** This wrapper is no longer thin.

File: cli/autostart_internal_test.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+		{
+			name:          "WhoKnows",
+			input:         []string{"Time", "is", "a", "human", "construct"},
+			expectedError: errInvalidTimeFormat.Error(),
+		},
+	} {
+		testCase := testCase
+		//nolint:paralleltest // t.Setenv
+		t.Run(testCase.name, func(t *testing.T) {
+			t.Setenv("TZ", testCase.tzEnv)
```
Comment (johnstcn) [inline cli/autostart_internal_test.go:94]: **Review:** Doing `setenv` here to validate that the timezone specified in the schedule is used

File: cli/autostart_internal_test.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+		{
+			name:          "DayOfWeekAndInvalidTimeAndLocation",
+			input:         []string{"nine", "Sun-Sat", "America/Chicago"},
+			expectedError: errInvalidTimeFormat.Error(),
+		},
+		{
+			name:          "TimezoneProvidedInsteadOfLocation",
+			input:         []string{"09:00AM", "Sun-Sat", "CST"},
+			expectedError: errUnsupportedTimezone.Error(),
+		},
```
Comment (johnstcn) [inline cli/autostart_internal_test.go:84]: @Emyrk this one just for you :-)

File: cli/autostart.go [hunk 6]
```diff
@@ -22,12 +31,12 @@ func autostart() *cobra.Command {
 		Use:         "autostart enable <workspace>",
 		Short:       "schedule a workspace to automatically start at a regular time",
 		Long:        autostartDescriptionLong,
-		Example:     "coder autostart enable my-workspace --minute 30 --hour 9 --days 1-5 --tz Europe/Dublin",
+		Example:     "coder autostart set my-workspace Mon-Fri 9:30AM Europe/Dublin",
```
Comment (johnstcn) [inline cli/autostart.go:34]:
```suggestion
		Example:     "coder autostart set my-workspace 9:30AM Mon-Fri Europe/Dublin",
```

**Review by mafredri — APPROVED** [review]
A few suggestions and questions, but in general looks good! 👍🏻

File: coderd/util/tz/tz_windows.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+	}
+
+	done := make(chan struct{})
+	go func() {
+		defer stdin.Close()
+		defer close(done)
+		_, _ = fmt.Fprintln(stdin, cmdTimezone)
+	}()
+
+	<-done
```
[the remark this answers is not in the document]
Comment (johnstcn) [reply coderd/util/tz/tz_windows.go:52]: I completely forgot about `strings.Reader`!

File: cli/autostart.go [hunk 8]
```diff
@@ -2,18 +2,27 @@ package cli
...
-When enabling autostart, provide the minute, hour, and day(s) of week.
-The default schedule is at 09:00 in your local timezone (TZ env, UTC by default).
+When enabling autostart, enter a schedule in the format: start-time [day-of-week] [location].
+  * Start-time (required) is accepted either in 12-hour (hh:mm{am|pm}) format, or 24-hour format hh:mm.
+  * Day-of-week (optional) allows specifying in the cron format, e.g. 1,3,5 or Mon-Fri.
+    Aliases such as @daily are not supported.
+    Default: * (every day)
+  * Location (optional) must be a valid location in the IANA timezone database.
+    If omitted, we will fall back to either the TZ environment variable or /etc/localtime.
+    You can check your corresponding location by visiting https://ipinfo.io - it shows in the demo widget on the right.
```
[the remark this answers is not in the document]
Comment (ammario) [reply cli/autostart.go:25]: Yes, this is excellent.

File: cli/autostart.go [hunk 9]
```diff
@@ -102,32 +104,30 @@ func autostartEnable() *cobra.Command {
...
 				return err
 			}
 
-			_, _ = fmt.Fprintf(cmd.OutOrStdout(), "\nThe %s workspace will automatically start at %s.\n\n", workspace.Name, validSchedule.Next(time.Now()))
-
+			schedNext := sched.Next(time.Now())
+			_, _ = fmt.Fprintf(cmd.OutOrStdout(),
+				"%s will automatically start at %s %s (%s)\n",
+				workspace.Name,
+				schedNext.In(sched.Location()).Format(time.Kitchen),
```
[the remark this answers is not in the document]
Comment (johnstcn) [reply cli/autostart.go:117]: This sounds like an internationalization problem to me, which I'm going to place inside a tin can and kick down the road for later. 😅

File: cli/autostart.go [hunk 10]
```diff
@@ -147,9 +147,95 @@ func autostartDisable() *cobra.Command {
...
+			_, _ = fmt.Fprintf(cmd.OutOrStdout(), "%s will no longer automatically start.\n", workspace.Name)
 
 			return nil
 		},
 	}
 }
+
+var errInvalidScheduleFormat = xerrors.New("Schedule must be in the format Mon-Fri 09:00AM America/Chicago")
+var errInvalidTimeFormat = xerrors.New("Start time must be in the format hh:mm[am|pm] or HH:MM")
+var errUnsupportedTimezone = xerrors.New("The location you provided looks like a timezone. Check https://ipinfo.io for your location.")
```
[the remark this answers is not in the document]
Comment (johnstcn) [reply cli/autostart.go:159]: We have something very similar inside another codebase which shall not be named ;-)
But yes, this needs more noodling.

**Review by ammario — APPROVED** [review]

**Pull request merged**